File: src/modules/users/infrastructure/persistence/user_repository.py
from fastapi import Depends
from typing import Annotated
from src.modules.users.ports.output import IUserRepository
from src.modules.shared.domain.entities import UserEntity
from src.modules.shared.infrastructure.models import UserModel
from src.libs.database.session import SessionDep


class UserRepository(IUserRepository):
    def __init__(self, session: SessionDep):
        self.session = session

    def create(self, user: UserEntity) -> UserEntity:
        model = self.__to_model(user)
        self.session.add(model)
        self.session.flush()
        return self.__to_entity(model)

    def find_by_id(self, user_id: str) -> UserEntity | None:
        model = self.session.get(UserModel, user_id)
        if model is None:
            return None
        return self.__to_entity(model)

    def __to_entity(self, model: UserModel) -> UserEntity:
        return UserEntity(id=model.id, username=model.username, email=model.email, is_active=model.is_active)

    def __to_model(self, entity: UserEntity) -> UserModel:
        return UserModel(id=entity.id, username=entity.username, email=entity.email, is_active=entity.is_active)

IUserRepoDep = Annotated[IUserRepository, Depends(UserRepository)]

File: src/modules/users/infrastructure/persistence/user.repository.ts
import { Inject, Injectable } from '@nestjs/common';
import { eq } from 'drizzle-orm';
import { DRIZZLE } from '../../../../libs/database/database.module';
import type {
  DrizzleClient,
  DrizzleDb,
} from '../../../../libs/database/drizzle-client';
import { getDb } from '../../../../libs/database/transaction.context';
import { usersTable } from '../../../../libs/database/schema';
import { UserEntity } from '../../../shared/domain/entities/user.entity';
import { IUserRepository } from '../../ports/output/user-repository.port';

type UserRow = typeof usersTable.$inferSelect;

@Injectable()
export class UserRepository implements IUserRepository {
  constructor(@Inject(DRIZZLE) private readonly db: DrizzleDb) {}

  private get client(): DrizzleClient {
    return getDb(this.db);
  }

  async create(user: UserEntity): Promise<UserEntity> {
    const [row] = await this.client
      .insert(usersTable)
      .values({
        username: user.username,
        email: user.email,
        isActive: user.isActive,
      })
      .returning();
    return this.toEntity(row);
  }

  async findById(id: string): Promise<UserEntity | null> {
    const [row] = await this.client
      .select()
      .from(usersTable)
      .where(eq(usersTable.id, id))
      .limit(1);
    return row ? this.toEntity(row) : null;
  }

  private toEntity(row: UserRow): UserEntity {
    return new UserEntity({
      id: row.id,
      username: row.username,
      email: row.email,
      isActive: row.isActive,
    });
  }
}
